Add AvroConverter in new copycat-avro-converter jar to convert Copycat and Avro data.

ewencp opened the pull request.

ConfluentJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.confluent.io/job/schema-registry-pr/33/<h2>Build result: FAILURE</span></h2>[...truncated 1119 lines...][INFO] Final Memory: 23M/89M[INFO] ------------------------------------------------------------------------[ERROR] Failed to execute goal on project kafka-avro-serializer: Could not resolve dependencies for project io.confluent:kafka-avro-serializer:jar:2.0-SNAPSHOT: Could not find artifact org.apache.kafka:kafka_2.10:jar:0.8.3-SNAPSHOT in jenkins (http://jenkins.confluent.io/plugin/repository/project/rest-utils/Build/19/repositoryChain) -> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <goals> -rf :kafka-avro-serializer[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/pom.xml to io.confluent/kafka-schema-registry-parent/2.0-SNAPSHOT/kafka-schema-registry-parent-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/target/kafka-schema-registry-parent-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-schema-registry-parent/2.0-SNAPSHOT/kafka-schema-registry-parent-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/package/pom.xml to io.confluent/kafka-schema-registry-package/2.0-SNAPSHOT/kafka-schema-registry-package-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/core/pom.xml to io.confluent/kafka-schema-registry/2.0-SNAPSHOT/kafka-schema-registry-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/pom.xml to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/target/kafka-schema-registry-client-2.0-SNAPSHOT.jar to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/target/kafka-schema-registry-client-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/json-serializer/pom.xml to io.confluent/kafka-json-serializer/2.0-SNAPSHOT/kafka-json-serializer-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-serializer/pom.xml to io.confluent/kafka-avro-serializer/2.0-SNAPSHOT/kafka-avro-serializer-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-converter/pom.xml to io.confluent/copycat-avro-converter/2.0-SNAPSHOT/copycat-avro-converter-2.0-SNAPSHOT.pomchannel stoppedSetting status of 861a3dc15941e3d8657b4d0848dd452a26869194 to FAILURE with url http://jenkins.confluent.io/job/schema-registry-pr/33/ and message: Build finished.
Test FAILed.

ewencp commented:
@gwenshap Would be good if you could review this since it's a lot like the Converter/JsonConverter code you just recently reviewed. Sorry about the extra noise, the first commit is from #219 and should go away once that's merged.

ConfluentJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.confluent.io/job/schema-registry-pr/34/<h2>Build result: FAILURE</span></h2>[...truncated 1123 lines...][INFO] ------------------------------------------------------------------------Waiting for Jenkins to finish collecting data[ERROR] Failed to execute goal on project kafka-avro-serializer: Could not resolve dependencies for project io.confluent:kafka-avro-serializer:jar:2.0-SNAPSHOT: Could not find artifact org.apache.kafka:kafka_2.10:jar:0.8.3-SNAPSHOT in jenkins (http://jenkins.confluent.io/plugin/repository/project/rest-utils/Build/19/repositoryChain) -> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <goals> -rf :kafka-avro-serializer[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/pom.xml to io.confluent/kafka-schema-registry-parent/2.0-SNAPSHOT/kafka-schema-registry-parent-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/target/kafka-schema-registry-parent-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-schema-registry-parent/2.0-SNAPSHOT/kafka-schema-registry-parent-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/package/pom.xml to io.confluent/kafka-schema-registry-package/2.0-SNAPSHOT/kafka-schema-registry-package-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/core/pom.xml to io.confluent/kafka-schema-registry/2.0-SNAPSHOT/kafka-schema-registry-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/pom.xml to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/target/kafka-schema-registry-client-2.0-SNAPSHOT.jar to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/target/kafka-schema-registry-client-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/json-serializer/pom.xml to io.confluent/kafka-json-serializer/2.0-SNAPSHOT/kafka-json-serializer-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-serializer/pom.xml to io.confluent/kafka-avro-serializer/2.0-SNAPSHOT/kafka-avro-serializer-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-converter/pom.xml to io.confluent/copycat-avro-converter/2.0-SNAPSHOT/copycat-avro-converter-2.0-SNAPSHOT.pomchannel stoppedSetting status of 861a3dc15941e3d8657b4d0848dd452a26869194 to FAILURE with url http://jenkins.confluent.io/job/schema-registry-pr/34/ and message: Build finished.
Test FAILed.

ConfluentJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.confluent.io/job/schema-registry-pr/35/<h2>Build result: FAILURE</span></h2>[...truncated 1098 lines...][INFO] ------------------------------------------------------------------------Waiting for Jenkins to finish collecting data[ERROR] Failed to execute goal on project kafka-avro-serializer: Could not resolve dependencies for project io.confluent:kafka-avro-serializer:jar:2.0-SNAPSHOT: Failed to collect dependencies for [org.apache.kafka:kafka_2.10:jar:0.8.3-SNAPSHOT (provided), org.apache.avro:avro:jar:1.7.7 (compile), io.confluent:kafka-schema-registry-client:jar:2.0-SNAPSHOT (compile), io.confluent:common-config:jar:2.0-SNAPSHOT (compile), junit:junit:jar:4.12 (test)]: Failed to read artifact descriptor for org.apache.kafka:kafka_2.10:jar:0.8.3-SNAPSHOT: Could not transfer artifact org.apache.kafka:kafka_2.10:pom:0.8.3-SNAPSHOT from/to ancestor (/home/jenkins/workspace/schema-registry-pr/maven/): No connector available to access repository ancestor (/home/jenkins/workspace/schema-registry-pr/maven/) of type default using the available factories WagonRepositoryConnectorFactory -> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <goals> -rf :kafka-avro-serializer[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/pom.xml to io.confluent/kafka-schema-registry-parent/2.0-SNAPSHOT/kafka-schema-registry-parent-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/target/kafka-schema-registry-parent-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-schema-registry-parent/2.0-SNAPSHOT/kafka-schema-registry-parent-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/package/pom.xml to io.confluent/kafka-schema-registry-package/2.0-SNAPSHOT/kafka-schema-registry-package-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/core/pom.xml to io.confluent/kafka-schema-registry/2.0-SNAPSHOT/kafka-schema-registry-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/pom.xml to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/target/kafka-schema-registry-client-2.0-SNAPSHOT.jar to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/target/kafka-schema-registry-client-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/json-serializer/pom.xml to io.confluent/kafka-json-serializer/2.0-SNAPSHOT/kafka-json-serializer-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-serializer/pom.xml to io.confluent/kafka-avro-serializer/2.0-SNAPSHOT/kafka-avro-serializer-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-converter/pom.xml to io.confluent/copycat-avro-converter/2.0-SNAPSHOT/copycat-avro-converter-2.0-SNAPSHOT.pomchannel stoppedSetting status of 861a3dc15941e3d8657b4d0848dd452a26869194 to FAILURE with url http://jenkins.confluent.io/job/schema-registry-pr/35/ and message: Build finished.
Test FAILed.

ConfluentJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.confluent.io/job/schema-registry-pr/36/<h2>Build result: FAILURE</span></h2>[...truncated 1098 lines...][INFO] ------------------------------------------------------------------------Waiting for Jenkins to finish collecting data[ERROR] Failed to execute goal on project kafka-avro-serializer: Could not resolve dependencies for project io.confluent:kafka-avro-serializer:jar:2.0-SNAPSHOT: Failed to collect dependencies for [org.apache.kafka:kafka_2.10:jar:0.8.3-SNAPSHOT (provided), org.apache.avro:avro:jar:1.7.7 (compile), io.confluent:kafka-schema-registry-client:jar:2.0-SNAPSHOT (compile), io.confluent:common-config:jar:2.0-SNAPSHOT (compile), junit:junit:jar:4.12 (test)]: Failed to read artifact descriptor for org.apache.kafka:kafka_2.10:jar:0.8.3-SNAPSHOT: Could not transfer artifact org.apache.kafka:kafka_2.10:pom:0.8.3-SNAPSHOT from/to ancestor (/home/jenkins/workspace/schema-registry-pr/maven/): No connector available to access repository ancestor (/home/jenkins/workspace/schema-registry-pr/maven/) of type default using the available factories WagonRepositoryConnectorFactory -> [Help 1][ERROR] [ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.[ERROR] Re-run Maven using the -X switch to enable full debug logging.[ERROR] [ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <goals> -rf :kafka-avro-serializer[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/pom.xml to io.confluent/kafka-schema-registry-parent/2.0-SNAPSHOT/kafka-schema-registry-parent-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/target/kafka-schema-registry-parent-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-schema-registry-parent/2.0-SNAPSHOT/kafka-schema-registry-parent-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/package/pom.xml to io.confluent/kafka-schema-registry-package/2.0-SNAPSHOT/kafka-schema-registry-package-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/core/pom.xml to io.confluent/kafka-schema-registry/2.0-SNAPSHOT/kafka-schema-registry-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/pom.xml to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/target/kafka-schema-registry-client-2.0-SNAPSHOT.jar to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/target/kafka-schema-registry-client-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/json-serializer/pom.xml to io.confluent/kafka-json-serializer/2.0-SNAPSHOT/kafka-json-serializer-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-serializer/pom.xml to io.confluent/kafka-avro-serializer/2.0-SNAPSHOT/kafka-avro-serializer-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-converter/pom.xml to io.confluent/copycat-avro-converter/2.0-SNAPSHOT/copycat-avro-converter-2.0-SNAPSHOT.pomchannel stoppedSetting status of 861a3dc15941e3d8657b4d0848dd452a26869194 to FAILURE with url http://jenkins.confluent.io/job/schema-registry-pr/36/ and message: Build finished.
Test FAILed.

ConfluentJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.confluent.io/job/schema-registry-pr/37/<h2>Build result: FAILURE</span></h2>[...truncated 6029 lines...][ERROR] For more information about the errors and possible solutions, please read the following articles:[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException[ERROR] [ERROR] After correcting the problems, you can resume the build with the command[ERROR]   mvn <goals> -rf :copycat-avro-converter[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/pom.xml to io.confluent/kafka-schema-registry-parent/2.0-SNAPSHOT/kafka-schema-registry-parent-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/target/kafka-schema-registry-parent-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-schema-registry-parent/2.0-SNAPSHOT/kafka-schema-registry-parent-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/package/pom.xml to io.confluent/kafka-schema-registry-package/2.0-SNAPSHOT/kafka-schema-registry-package-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/core/pom.xml to io.confluent/kafka-schema-registry/2.0-SNAPSHOT/kafka-schema-registry-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/core/target/kafka-schema-registry-2.0-SNAPSHOT.jar to io.confluent/kafka-schema-registry/2.0-SNAPSHOT/kafka-schema-registry-2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/core/target/kafka-schema-registry-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-schema-registry/2.0-SNAPSHOT/kafka-schema-registry-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/pom.xml to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/target/kafka-schema-registry-client-2.0-SNAPSHOT.jar to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/client/target/kafka-schema-registry-client-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-schema-registry-client/2.0-SNAPSHOT/kafka-schema-registry-client-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/json-serializer/pom.xml to io.confluent/kafka-json-serializer/2.0-SNAPSHOT/kafka-json-serializer-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/json-serializer/target/kafka-json-serializer-2.0-SNAPSHOT.jar to io.confluent/kafka-json-serializer/2.0-SNAPSHOT/kafka-json-serializer-2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/json-serializer/target/kafka-json-serializer-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-json-serializer/2.0-SNAPSHOT/kafka-json-serializer-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-serializer/pom.xml to io.confluent/kafka-avro-serializer/2.0-SNAPSHOT/kafka-avro-serializer-2.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-serializer/target/kafka-avro-serializer-2.0-SNAPSHOT.jar to io.confluent/kafka-avro-serializer/2.0-SNAPSHOT/kafka-avro-serializer-2.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-serializer/target/kafka-avro-serializer-2.0-SNAPSHOT-tests.jar to io.confluent/kafka-avro-serializer/2.0-SNAPSHOT/kafka-avro-serializer-2.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/schema-registry-pr/avro-converter/pom.xml to io.confluent/copycat-avro-converter/2.0-SNAPSHOT/copycat-avro-converter-2.0-SNAPSHOT.pomchannel stoppedSetting status of 861a3dc15941e3d8657b4d0848dd452a26869194 to FAILURE with url http://jenkins.confluent.io/job/schema-registry-pr/37/ and message: Build finished.
Test FAILed.

ConfluentJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.confluent.io/job/schema-registry-pr/39/<h2>Failed Tests: <span class='status-failure'>2</span></h2><h3><a name='schema-registry-pr/io.confluent:kafka-schema-registry' /><a href='http://jenkins.confluent.io/job/schema-registry-pr/39/io.confluent$kafka-schema-registry/testReport'>schema-registry-pr/io.confluent:kafka-schema-registry</a>: <span class='status-failure'>2</span></h3><ul><li><a href='http://jenkins.confluent.io/job/schema-registry-pr/39/io.confluent$kafka-schema-registry/testReport/io.confluent.kafka.schemaregistry.rest/RestApiTest/testLookUpNonExistentSchemaUnderSubject/'><strong>io.confluent.kafka.schemaregistry.rest.RestApiTest.testLookUpNonExistentSchemaUnderSubject</strong></a></li><li><a href='http://jenkins.confluent.io/job/schema-registry-pr/39/io.confluent$kafka-schema-registry/testReport/io.confluent.kafka.schemaregistry.zookeeper/MasterElectorTest/testSlaveIsNeverMaster/'><strong>io.confluent.kafka.schemaregistry.zookeeper.MasterElectorTest.testSlaveIsNeverMaster</strong></a></li></ul>
Test FAILed.

ewencp commented:
retest this please

ConfluentJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.confluent.io/job/schema-registry-pr/41/
Test PASSed.

gwenshap commented:
clearly being polite to Jenkins helps.

gwenshap commented on avro-converter/src/main/java/io/confluent/copycat/avro/AvroConverter.java:
```diff
@@ -0,0 +1,107 @@
+    deserializer = new Deserializer();
+  }
+
+  public AvroConverter(SchemaRegistryClient client) {
```
Is this used just for testing? if so, maybe clarify in a comment (or an annotation? I vaguely remember an annotation to that effect)

ewencp replied:
Guava has an annotation for this (VisibleForTesting), but I don't think there's one built-in or in any of the deps we already use. I've just used EasyMock/PowerMock in other code (e.g. the Copycat packages in Kafka) to avoid this messiness. I stuck with the current convention here, but we could also just start using that approach in this code.

gwenshap commented on avro-converter/src/main/java/io/confluent/copycat/avro/AvroConverter.java:
```diff
@@ -0,0 +1,107 @@
+
+  @Override
+  public void configure(Map<String, ?> configs, boolean isKey) {
+    this.isKey = isKey;
```
what about configs?

gwenshap replied:
Also, is configure tested?

ewencp replied:
Good catch, this code wasn't actually complete. Added a test for configure, ended up adding powermock for it since otherwise it's difficult to have any assertions after calling `configure()` without it.

gwenshap commented on avro-converter/src/main/java/io/confluent/copycat/avro/AvroData.java:
```diff
@@ -0,0 +1,1073 @@
+  // Avro does not permit empty schema names, which might be the ideal default since we also are
+  // not permitted to simply omit the name. Instead, make it very clear where the default is
+  // coming from.
+  public static final String DEFAULT_SCHEMA_NAME = "CopycatDefault";
```
Would it make sense to just force Copycat Schema to have a non-empty name? 
I can see why we may not want to force to much Avro-isms on Copycat data, but schema name isn't too convoluted.

ewencp replied:
One major difference is that in Copycat I let any schema be named whereas Avro has specific types that can be named (record, enum, etc). It's annoying to force all to be named because then even primitive types need to be named and all field schemas need names in addition to the field name. I think forcing naming is likely to just be confusing anyway since a lot of Copycat schemas are autogenerated and so the naming wouldn't necessarily be consistent. My primary intent with names was to provide a mechanism for identifying logical types to allow for special handling.

gwenshap commented on avro-converter/src/main/java/io/confluent/copycat/avro/AvroData.java:
```diff
@@ -0,0 +1,1073 @@
+
+  public static final String COPYCAT_NAME_PROP = "copycat_name";
+  public static final String COPYCAT_DOC_PROP = "copycat_doc";
+  public static final String COPYCAT_VERSION_PROP = "copycat_version";
```
underscores?  we are using dotted properties everywhere else, right?

ewencp replied:
Huh, I think this was just auto-pilot where I avoided using dots for field names since it's valid but requires you to use json["copycat.name"] style syntax since json.copycat.name would refer to something else. I actually don't like this since I think the `_` is ugly.

There are a couple of options here. We could switch to dots as you say. I could also potentially rework this to put all the extra copycat info into a single `copycat` field that is an object with it's own fields like `name`, `doc`, `version`.

gwenshap replied:
I don't have strong preference either way.

gwenshap commented on avro-converter/src/main/java/io/confluent/copycat/avro/AvroData.java:
```diff
@@ -0,0 +1,1073 @@
+
+  public static final String AVRO_TYPE_ANYTHING = NAMESPACE + ".Anything";
+
+  private static final Map<String, Schema.Type> SPECIALIZED_TYPES_BY_TYPE_CODE = new HashMap<>();
```
Why do we have specialized types? Aren't these just short and byte and should be treated like other types?

gwenshap replied:
Ah! Those don't exist in Avro :)
NON_AVRO_TYPES is an ugly name, but probably clearer...

ewencp replied:
Changed.

gwenshap commented on avro-converter/src/main/java/io/confluent/copycat/avro/AvroData.java:
```diff
@@ -0,0 +1,1073 @@
+    // Intuitively this should be a union schema. However, unions can't be named in Avro and this
+    // is a self-referencing type, so we need to use a format in which we can name the entire schema
+
+    ANYTHING_SCHEMA =
```
Specialized types are not part of anything? I feel like there's a good idea here that I'm somehow missing...

gwenshap commented on avro-converter/src/test/java/io/confluent/copycat/avro/AvroDataTest.java:
```diff
@@ -0,0 +1,671 @@
+
+  }
+
+  private static NonRecordContainer assertNonRecordConversion(
```
Why return the container? Asserts are typically void.

ewencp replied:
Only because the assert does the conversion and in one case I want to do one additional check. I've changed the name to `checkNonRecordConversion` so it is not misleading.

gwenshap commented on src/assembly/package.xml:
```diff
@@ -56,6 +59,16 @@
+            <scope>runtime</scope>
+            <useTransitiveFiltering>true</useTransitiveFiltering>
+            <includes>
+                <include>io.confluent:kafka-json-serializer</include>
```
I thought the json serializer is part of Kafka itself. Should this be kafka-avro-serializer?

ewencp replied:
We actually added a JSON serializer to schema registry before, I just noticed this was missing as I was adding the converter. I can move it out into a separate patch if you prefer.

gwenshap replied:
its fine.

gwenshap commented:
Looks good (although I admit I expect this to be a lot simpler given that Copycat data was supposed to be fairly similar to Avro). I just had few questions / comments.

ConfluentJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.confluent.io/job/schema-registry-pr/42/
Test PASSed.

ewencp commented:
@gwenshap Addressed all your comments except the one about underscores. To be honest, that's just a naming preference thing and I'm happy to go with whatever everyone else things is best.

I also added some scripts/configs -- I realized there wasn't a good way of testing that this actually works. That also affected packaging scripts a bit because I needed to make the copycat runtime jars easily available to those scripts. Now you should be able to fire up the appropriate dependent services, run using Kafka's config/copycat-console-source.properties connector config, type a few lines, then see things appear in Kafka by consuming them with the kafka-avro-console-consumer. Only validates strings, but at least it's easier to check that it's really working now.

ewencp commented:
By the way, re: size of the patch -- the converters were surprisingly large SLOC-wise even when the APIs matched *exactly*, i.e. when I was using a hacked up Avro for the data API. I think this just boils down to the impedance mismatch between a fluent API for schema and record construction, which is what user's *normally* want, and the API that you'd want for something like Copycat where a lot of translation is automated.

A really simple example of this is how you construct schemas of a specific type. In a fluent API, you want something like `SchemaBuilder.int32().optional().build()`, but in an automated translation you want something more like `SchemaBuilder.type(INT32).setOptional(true).build()` because you might have a couple of switch or if/else statements that determine the schema "flags", but then are able to construct the full schema in a single statement.

I wrote up the Copycat API thinking mostly of connector devs because the API will have a lot more users coming from that perspective and only a rare few having to implement convertilizers. That said, if there's something we could add/modify in the Copycat data API to make this less messy, I'm all for it -- right now I'm the one feeling most of the pain from having relatively verbose code for performing this translation.

gwenshap commented on avro-converter/src/test/java/io/confluent/copycat/avro/AvroDataTest.java:
```diff
@@ -0,0 +1,671 @@
+
+import static org.junit.Assert.*;
+
+public class AvroDataTest {
```
I don't think we are checking any failed conversions... shouldn't we validate that the expected exception is thrown when conversion fails?

gwenshap commented:
LGTM. Up to you regarding the dots vs underscores, and whether the failed conversion tests are actually missing. Thanks for making testing possible. I assume Muckrake tests will follow?

ConfluentJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.confluent.io/job/schema-registry-pr/43/
Test PASSed.

ewencp commented:
@gwenshap Good catch on missing failed conversions, I added some and caught a few issues. Yes, system tests are planned. Need to get kafkatest published and figure out our strategy for pinning to development versions since I want to reuse at least the existing Copycat service for tests since we'll want the same basic set of tests as the JsonConverter.
